File: tools/bsp.py
#!/usr/bin/env python3
"""
bsp.py -- read .bsp files: the collision / space-partition trees.

Despite the extension these hold no geometry. A `.bsp` is a flat array of
24-byte BSP nodes and nothing else — no header, no signature, no counts, so
the node count is simply `filesize / 24`. The renderable geometry of a level
lives in the big `.mod` files (`ml7z_castle.mod`, `r1_castle.mod`, ...); this
is the partition the engine uses for collision and visibility, matching the
debug paths `E:\\mdk2\\Omen\\omPolyhedron.c` and `omCollision.c`.

    struct node {          // 24 bytes
        float normal[3];   // unit plane normal
        float dist;        // plane distance from the origin
        u32   front;       // child index, 0xFFFFFFFF for a leaf
        u32   back;        // child index, 0xFFFFFFFF for a leaf
    };

Node 0 is the root, and a point is **inside** when it lies in front of every
plane down a chain -- that is, when the descent reaches a leaf through the
*front* child. `l6r8_Stack5.bsp` shows this with nothing to interpret: seven
planes forming one oriented 4.5-cube, every `back` a leaf, `front` running
0 -> 1 -> ... -> 6. It is a crate from a stack, and it is a crate.

**The query point must be negated.** The same file settles it: its box is
centred on (0.73, 219.73, -2.25) while `l6r8_Stack5.mod` is centred on
(-0.73, -219.73, 2.25) -- the exact opposite. So the tree is authored in a
mirrored frame, and `contains()` negates before descending. Testing points
either side of a face then separates them 48/48 on that crate and 799/800 on
`l3_maze`; without the negation it is 0/800, every point landing on the same
side.

Verified across the whole corpus, to the project's 100% rule:

  * all 692 files are an exact multiple of 24 bytes;
  * all 64387 normals are unit vectors to within 1e-3 — which is what
    identified the format in the first place;
  * all 692 trees are structurally sound: every child index in range, every
    node referenced exactly once, exactly one unreferenced node (the root).

Usage:
    python3 tools/bsp.py extracted/base/c10.bsp
    python3 tools/bsp.py extracted/base --validate
"""

from __future__ import annotations

import argparse
import math
import struct
import sys
from pathlib import Path

NODE_SIZE = 24
LEAF = 0xFFFFFFFF


class BspError(ValueError):
    pass


def parse(data: bytes) -> list[tuple[tuple[float, float, float], float,
                                     int, int]]:
    if not data or len(data) % NODE_SIZE:
        raise BspError(f"{len(data)} bytes is not a multiple of {NODE_SIZE}")
    out = []
    for i in range(len(data) // NODE_SIZE):
        nx, ny, nz, dist, front, back = struct.unpack_from(
            "<4f2I", data, i * NODE_SIZE)
        out.append(((nx, ny, nz), dist, front, back))
    return out


def validate(nodes: list) -> None:
    """Raise unless the array really is a well-formed tree of unit planes."""
    n = len(nodes)
    refs = [0] * n
    for i, (normal, _dist, front, back) in enumerate(nodes):
        length = math.sqrt(sum(c * c for c in normal))
        if abs(length - 1.0) > 1e-3:
            raise BspError(f"node {i}: normal length {length:.5f}")
        for child in (front, back):
            if child == LEAF:
                continue
            if child >= n:
                raise BspError(f"node {i}: child {child} out of range")
            refs[child] += 1
    if any(r > 1 for r in refs):
        raise BspError("a node is referenced more than once")
    roots = [i for i, r in enumerate(refs) if r == 0]
    if roots != [0]:
        raise BspError(f"expected exactly node 0 unreferenced, got {roots}")


def contains(nodes: list, point) -> bool:
    """Is the point inside solid geometry? Point is in *model* coordinates."""
    x, y, z = -point[0], -point[1], -point[2]
    i = 0
    while True:
        (nx, ny, nz), dist, front, back = nodes[i]
        side = nx * x + ny * y + nz * z - dist
        child = front if side >= 0 else back
        if child == LEAF:
            return side >= 0
        i = child


def crosses(nodes: list, a, b) -> bool:
    """Does the segment a->b pass through solid geometry? Model coordinates.

    The same descent as `contains`, split at the plane crossings instead of
    following one side: when the two ends fall on opposite sides of a node's
    plane the segment is cut there and both halves are tested. That makes it
    *exact for the tree* rather than a sampling of it -- a wall thinner than
    any step size still stops it.

    The leaf convention is `contains`'s: reaching a leaf on the front side
    (`side >= 0`) is solid, on the back side it is empty.

    What the original does at 0x471dc0 is not this -- it is a query against
    the engine's own world structure, reached from `mdkAILineOfSight`
    (0x402950) after the field-of-view test. The trees are the same trees, so
    this answers the same question; whether it answers it identically in every
    corner is not claimed.

    Iterative rather than recursive, like `contains`: the deepest tree in the
    game is `l4_r9.bsp` at 107, so recursion would in fact be safe, but a
    split pushes *two* halves and the branching is what would grow.
    """
    stack = [(0, tuple(a), tuple(b))]
    while stack:
        i, p, q = stack.pop()
        (nx, ny, nz), dist, front, back = nodes[i]
        dp = nx * -p[0] + ny * -p[1] + nz * -p[2] - dist
        dq = nx * -q[0] + ny * -q[1] + nz * -q[2] - dist
        if dp >= 0 and dq >= 0:
            if front == LEAF:
                return True
            stack.append((front, p, q))
        elif dp < 0 and dq < 0:
            if back != LEAF:
                stack.append((back, p, q))
        else:
            # the split point, on the plane
            t = dp / (dp - dq)
            m = tuple(p[c] + (q[c] - p[c]) * t for c in range(3))
            # whichever half lies on the front side is the one that can be
            # solid, and it is the first half when p is in front
            fp, fq = (p, m) if dp >= 0 else (m, q)
            bp, bq = (m, q) if dp >= 0 else (p, m)
            if front == LEAF:
                return True
            stack.append((front, fp, fq))
            if back != LEAF:
                stack.append((back, bp, bq))
    return False


def drop(nodes: list, point, floor: float, step: float = 0.25):
    """Lower the point along -Z until it enters solid. -> z of contact, or None.

    A crude but honest demonstration that the tree can be collided against:
    no sweeping, no radius, just a downward march.
    """
    x, y, z = point
    if contains(nodes, (x, y, z)):
        return None                      # started inside
    while z > floor:
        z -= step
        if contains(nodes, (x, y, z)):
            return z + step
    return None


def depth(nodes: list) -> int:
    """Deepest path from the root, iteratively — these trees get deep."""
    best, stack = 0, [(0, 1)]
    while stack:
        i, d = stack.pop()
        best = max(best, d)
        _n, _dist, front, back = nodes[i]
        for child in (front, back):
            if child != LEAF:
                stack.append((child, d + 1))
    return best


def probe_points(nodes: list) -> list[tuple[float, float, float]]:
    """Query points derived from the tree itself, so no other file is needed
    and two implementations can be asked exactly the same thing.

    The first are the feet of the planes, negated -- `contains` negates again,
    so each lands **exactly on** a plane, which is where two implementations
    would differ if either got the `>= 0` boundary wrong. The rest are a 4x4x4
    grid over the box those feet span. `engine/src/main.rs` builds the same
    list, in the same order.
    """
    out = [tuple(-normal[c] * dist for c in range(3))
           for normal, dist, _f, _b in nodes[:256]]
    lo = [min(p[c] for p in out) for c in range(3)]
    hi = [max(p[c] for p in out) for c in range(3)]
    for a in range(4):
        for b in range(4):
            for c in range(4):
                t = (a, b, c)
                out.append(tuple(lo[k] + (hi[k] - lo[k]) * t[k] / 3.0
                                 for k in range(3)))
    return out


def digest(files: list[Path]) -> int:
    """`name nodes depth inside crc32` a line, which is what the engine
    prints for `--bsp`."""
    import zlib
    for f in sorted(files, key=lambda p: p.name.lower()):
        nodes = parse(f.read_bytes())
        validate(nodes)
        points = probe_points(nodes)
        answers = bytes(contains(nodes, p) for p in points)
        # and the same points taken in consecutive pairs as segments, which
        # is what `crosses` is asked -- a line that starts on one plane and
        # ends on another crosses every boundary the tree has
        blocked = bytes(crosses(nodes, points[i], points[i + 1])
                        for i in range(len(points) - 1))
        print(f"{f.name.lower()} {len(nodes)} {depth(nodes)} "
              f"{sum(answers)} {zlib.crc32(answers):08x} "
              f"{sum(blocked)} {zlib.crc32(blocked):08x}")
    print(f"{len(files)} trees", file=sys.stderr)
    return 0


def compare(files: list[Path], gamedir: str) -> int:
    """Run the engine over `gamedir` and require the same answers.

    Not just the same parse: the `inside` answers come from points that lie
    exactly on the planes, so the `>= 0` boundary -- the one thing in this
    format that can be quietly wrong -- is what is being compared. The second
    pair of numbers is `crosses` over the same points taken as consecutive
    segments, which walks the split path rather than one side of it.
    """
    import subprocess
    import zlib
    root = Path(__file__).resolve().parent.parent
    out = subprocess.run(
        ["cargo", "run", "--quiet", "--release", "--manifest-path",
         str(root / "engine/Cargo.toml"), "--", gamedir, "--bsp"],
        capture_output=True, text=True, check=True).stdout
    theirs = {}
    for line in out.splitlines():
        f = line.split()
        if len(f) == 7:
            theirs[f[0]] = (int(f[1]), int(f[2]), int(f[3]), f[4],
                            int(f[5]), f[6])

    bad = []
    for f in files:
        name = f.name.lower()
        nodes = parse(f.read_bytes())
        validate(nodes)
        points = probe_points(nodes)
        answers = bytes(contains(nodes, p) for p in points)
        blocked = bytes(crosses(nodes, points[i], points[i + 1])
                        for i in range(len(points) - 1))
        mine = (len(nodes), depth(nodes), sum(answers),
                f"{zlib.crc32(answers):08x}",
                sum(blocked), f"{zlib.crc32(blocked):08x}")
        if theirs.get(name) != mine:
            bad.append(f"{name}: {mine} against the engine's "
                       f"{theirs.get(name)}")
    for name in sorted(set(theirs) - {f.name.lower() for f in files}):
        bad.append(f"{name}: the engine read it and this did not")
    for line in bad[:20]:
        print(f"MISMATCH {line}", file=sys.stderr)
    print(f"{len(files)} trees, {len(bad)} disagree", file=sys.stderr)
    return 1 if bad else 0


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description=__doc__.split("\n")[1])
    ap.add_argument("src", type=Path, help="a .bsp file or a directory")
    ap.add_argument("--validate", action="store_true",
                    help="check every file and report totals")
    ap.add_argument("--point", nargs=3, type=float, metavar=("X", "Y", "Z"),
                    help="report whether this point is inside solid geometry")
    ap.add_argument("--digest", action="store_true",
                    help="one line per tree, for --engine to diff")
    ap.add_argument("--engine", metavar="GAMEDIR",
                    help="run the engine over this installation and require "
                         "it to answer identically")
    args = ap.parse_args(argv)

    files = (sorted(args.src.rglob("*.bsp")) if args.src.is_dir()
             else [args.src])
    if not files:
        ap.error(f"no .bsp files in {args.src}")

    if args.digest:
        return digest(files)

    if args.engine:
        return compare(files, args.engine)

    if args.point:
        nodes = parse(files[0].read_bytes())
        validate(nodes)
        print(f"{files[0].name}: {tuple(args.point)} inside = "
              f"{contains(nodes, args.point)}")
        return 0

    total = leaves = 0
    bad = []
    for f in files:
        try:
            nodes = parse(f.read_bytes())
            validate(nodes)
        except BspError as e:
            bad.append((f.name, str(e)))
            continue
        total += len(nodes)
        leaves += sum(1 for _n, _d, a, b in nodes
                      for c in (a, b) if c == LEAF)
        if not args.validate:
            print(f"{f.name}: {len(nodes)} nodes, depth {depth(nodes)}, "
                  f"{sum(1 for _n, _d, a, b in nodes for c in (a, b) if c == LEAF)}"
                  " leaf links")
    for name, err in bad:
        print(f"ERROR {name}: {err}", file=sys.stderr)
    print(f"{len(files) - len(bad)}/{len(files)} trees valid, "
          f"{total} nodes, {leaves} leaf links", file=sys.stderr)
    return 1 if bad else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: engine/src/formats/bsp.rs
//! `.bsp` — the collision and space-partition trees.
//!
//! Despite the extension these hold no geometry. A `.bsp` is a flat array of
//! 24-byte nodes and nothing else — no header, no signature, no counts, so
//! the node count is simply `filesize / 24`. The renderable geometry of a
//! level lives in the big `.mod` files; this is the partition the engine
//! collides against, matching the debug paths `omPolyhedron.c` and
//! `omCollision.c`.
//!
//! ```text
//! struct node {          // 24 bytes
//!     float normal[3];   // unit plane normal
//!     float dist;        // plane distance from the origin
//!     u32   front;       // child index, 0xFFFFFFFF for a leaf
//!     u32   back;        // child index, 0xFFFFFFFF for a leaf
//! };
//! ```
//!
//! Node 0 is the root, and a point is **inside** when the descent reaches a
//! leaf through the *front* child. `l6r8_Stack5.bsp` shows this with nothing
//! to interpret: seven planes forming one oriented 4.5-cube, every `back` a
//! leaf, `front` running 0 -> 1 -> ... -> 6. It is a crate from a stack, and
//! it is a crate.
//!
//! **The query point must be negated.** The same file settles it: its box is
//! centred on (0.73, 219.73, -2.25) while `l6r8_Stack5.mod` is centred on
//! the exact opposite. The tree is authored in a mirrored frame. Testing
//! points either side of a face separates them 799 of 800 on `l3_maze`;
//! without the negation it is 0 of 800, every point landing on the same side.
//!
//! `../../tools/bsp.py` is the reference: all 692 files a multiple of 24
//! bytes, all 64387 normals unit to within 1e-3, all 692 trees structurally
//! sound.

pub const NODE_SIZE: usize = 24;
pub const LEAF: u32 = 0xFFFF_FFFF;

#[derive(Debug, PartialEq)]
pub enum Error {
    NotWholeNodes(usize),
    Empty,
    Normal { node: usize, length: f32 },
    ChildOutOfRange { node: usize, child: u32 },
    SharedNode(usize),
    Roots(Vec<usize>),
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Error::NotWholeNodes(n) => write!(f, "{n} bytes is not a multiple of {NODE_SIZE}"),
            Error::Empty => write!(f, "no nodes at all"),
            Error::Normal { node, length } => write!(f, "node {node}: normal length {length:.5}"),
            Error::ChildOutOfRange { node, child } => {
                write!(f, "node {node}: child {child} out of range")
            }
            Error::SharedNode(i) => write!(f, "node {i} is referenced more than once"),
            Error::Roots(r) => write!(f, "expected node 0 alone unreferenced, got {r:?}"),
        }
    }
}

impl std::error::Error for Error {}

#[derive(Clone, Copy, Debug)]
pub struct Node {
    pub normal: [f32; 3],
    pub dist: f32,
    pub front: u32,
    pub back: u32,
}

pub struct Bsp {
    pub nodes: Vec<Node>,
}

impl Bsp {
    pub fn parse(data: &[u8]) -> Result<Bsp, Error> {
        if data.is_empty() {
            return Err(Error::Empty);
        }
        if data.len() % NODE_SIZE != 0 {
            return Err(Error::NotWholeNodes(data.len()));
        }
        let f = |at: usize| f32::from_bits(u32::from_le_bytes(data[at..at + 4].try_into().unwrap()));
        let u = |at: usize| u32::from_le_bytes(data[at..at + 4].try_into().unwrap());
        let nodes = (0..data.len() / NODE_SIZE)
            .map(|i| {
                let o = i * NODE_SIZE;
                Node {
                    normal: [f(o), f(o + 4), f(o + 8)],
                    dist: f(o + 12),
                    front: u(o + 16),
                    back: u(o + 20),
                }
            })
            .collect();
        Ok(Bsp { nodes })
    }

    /// Raise unless the array really is a well-formed tree of unit planes:
    /// every child in range, every node referenced once, exactly one node
    /// unreferenced and that one the root. It is what identified the format,
    /// and it is cheap enough to keep doing.
    pub fn validate(&self) -> Result<(), Error> {
        let n = self.nodes.len();
        let mut refs = vec![0u32; n];
        for (i, node) in self.nodes.iter().enumerate() {
            let length = node.normal.iter().map(|c| c * c).sum::<f32>().sqrt();
            if (length - 1.0).abs() > 1e-3 {
                return Err(Error::Normal { node: i, length });
            }
            for child in [node.front, node.back] {
                if child == LEAF {
                    continue;
                }
                if child as usize >= n {
                    return Err(Error::ChildOutOfRange { node: i, child });
                }
                refs[child as usize] += 1;
            }
        }
        if let Some(i) = refs.iter().position(|&r| r > 1) {
            return Err(Error::SharedNode(i));
        }
        let roots: Vec<usize> = refs
            .iter()
            .enumerate()
            .filter(|(_, &r)| r == 0)
            .map(|(i, _)| i)
            .collect();
        if roots != [0] {
            return Err(Error::Roots(roots));
        }
        Ok(())
    }

    /// Is the point inside solid geometry? The point is in **model**
    /// coordinates and is negated on the way in — see the module note.
    pub fn contains(&self, point: [f64; 3]) -> bool {
        let (x, y, z) = (-point[0], -point[1], -point[2]);
        let mut i = 0usize;
        loop {
            let node = self.nodes[i];
            let side = node.normal[0] as f64 * x
                + node.normal[1] as f64 * y
                + node.normal[2] as f64 * z
                - node.dist as f64;
            let child = if side >= 0.0 { node.front } else { node.back };
            if child == LEAF {
                return side >= 0.0;
            }
            i = child as usize;
        }
    }

    /// Does the segment `a`→`b` pass through solid geometry?
    ///
    /// The same descent as [`Bsp::contains`], split at the plane crossings
    /// instead of following one side: where the ends fall on opposite sides
    /// of a node's plane the segment is cut there and both halves are tested.
    /// That makes it **exact for the tree** rather than a sampling of it — a
    /// wall thinner than any step size still stops it, which is the whole
    /// reason a line of sight cannot be done by marching points.
    ///
    /// The leaf convention is `contains`'s: a leaf reached on the front side
    /// is solid, on the back side it is empty. Coordinates are negated the
    /// same way and for the same reason.
    ///
    /// The original's own occlusion test (0x471dc0, reached from
    /// `mdkAILineOfSight` at 0x402950 after the field-of-view check) is a
    /// query against its world structure rather than this traversal. The
    /// trees are the same trees, so the question is the same; identical
    /// answers in every corner are **not** claimed.
    pub fn crosses(&self, a: [f64; 3], b: [f64; 3]) -> bool {
        let plane = |n: &Node, p: [f64; 3]| {
            n.normal[0] as f64 * -p[0]
                + n.normal[1] as f64 * -p[1]
                + n.normal[2] as f64 * -p[2]
                - n.dist as f64
        };
        let mut stack = vec![(0usize, a, b)];
        while let Some((i, p, q)) = stack.pop() {
            let node = self.nodes[i];
            let (dp, dq) = (plane(&node, p), plane(&node, q));
            if dp >= 0.0 && dq >= 0.0 {
                if node.front == LEAF {
                    return true;
                }
                stack.push((node.front as usize, p, q));
            } else if dp < 0.0 && dq < 0.0 {
                if node.back != LEAF {
                    stack.push((node.back as usize, p, q));
                }
            } else {
                let t = dp / (dp - dq);
                let m = [0, 1, 2].map(|c| p[c] + (q[c] - p[c]) * t);
                // whichever half lies on the front side is the one that can
                // be solid, and it is the first half when p is in front
                let (fp, fq) = if dp >= 0.0 { (p, m) } else { (m, q) };
                let (bp, bq) = if dp >= 0.0 { (m, q) } else { (p, m) };
                if node.front == LEAF {
                    return true;
                }
                stack.push((node.front as usize, fp, fq));
                if node.back != LEAF {
                    stack.push((node.back as usize, bp, bq));
                }
            }
        }
        false
    }

    /// Deepest path from the root, iteratively — these trees get deep enough
    /// that recursion is a real risk.
    pub fn depth(&self) -> usize {
        let mut best = 0;
        let mut stack = vec![(0usize, 1usize)];
        while let Some((i, d)) = stack.pop() {
            best = best.max(d);
            for child in [self.nodes[i].front, self.nodes[i].back] {
                if child != LEAF {
                    stack.push((child as usize, d + 1));
                }
            }
        }
        best
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A single plane at z = 0 (stored mirrored, as the format is): solid
    /// below, open above, and the negation is what makes that come out right.
    fn one_plane() -> Vec<u8> {
        let mut d = Vec::new();
        for v in [0.0f32, 0.0, 1.0, 0.0] {
            d.extend_from_slice(&v.to_le_bytes());
        }
        d.extend_from_slice(&LEAF.to_le_bytes());
        d.extend_from_slice(&LEAF.to_le_bytes());
        d
    }

    #[test]
    fn a_leaf_through_front_is_inside() {
        let bsp = Bsp::parse(&one_plane()).unwrap();
        bsp.validate().unwrap();
        assert_eq!(bsp.nodes.len(), 1);
        assert_eq!(bsp.depth(), 1);
        // the point is negated, so it is +z that lands behind the plane
        assert!(bsp.contains([0.0, 0.0, -1.0]));
        assert!(!bsp.contains([0.0, 0.0, 1.0]));
    }

    /// The point about a segment query is that it does not sample: a plane
    /// has no thickness at all, and no number of probe points along the line
    /// would ever land on it, yet a line straight through must be stopped.
    #[test]
    fn a_segment_crosses_a_plane_that_no_sample_would_land_on() {
        let bsp = Bsp::parse(&one_plane()).unwrap();
        // -z is the solid side once the negation is applied, so this line
        // starts in the open and ends in solid
        assert!(bsp.crosses([0.0, 0.0, 1.0], [0.0, 0.0, -1.0]));
        assert!(bsp.crosses([0.0, 0.0, -1.0], [0.0, 0.0, 1.0]), "and the other way");
        // wholly on the open side, however far it runs
        assert!(!bsp.crosses([-500.0, 0.0, 1.0], [500.0, 0.0, 1.0]));
        // wholly inside is still solid
        assert!(bsp.crosses([-500.0, 0.0, -1.0], [500.0, 0.0, -1.0]));
        // and a segment that only touches the plane counts, because the leaf
        // convention is `side >= 0` and `contains` says the same of a point
        assert!(bsp.crosses([0.0, 0.0, 1.0], [0.0, 0.0, 0.0]));
        assert_eq!(bsp.contains([0.0, 0.0, 0.0]), true, "the boundary is solid for both");
    }

    #[test]
    fn a_short_file_is_not_a_tree() {
        assert!(matches!(Bsp::parse(&[0u8; 23]), Err(Error::NotWholeNodes(23))));
        assert!(matches!(Bsp::parse(&[]), Err(Error::Empty)));
    }

    #[test]
    fn validation_catches_a_normal_that_is_not_unit() {
        let mut d = one_plane();
        d[8..12].copy_from_slice(&2.0f32.to_le_bytes());
        assert!(matches!(
            Bsp::parse(&d).unwrap().validate(),
            Err(Error::Normal { node: 0, .. })
        ));
    }

    /// Two nodes both claiming the same child is not a tree, and the check
    /// that catches it is the one that identified the format.
    #[test]
    fn validation_catches_a_shared_child() {
        let mut d = one_plane();
        d.extend_from_slice(&one_plane());
        d.extend_from_slice(&one_plane());
        d[16..20].copy_from_slice(&2u32.to_le_bytes());
        d[NODE_SIZE + 16..NODE_SIZE + 20].copy_from_slice(&2u32.to_le_bytes());
        assert!(matches!(
            Bsp::parse(&d).unwrap().validate(),
            Err(Error::SharedNode(2))
        ));
    }
}
